netket/vqs/exact/exact_expect.py

```diff
@@ -0,0 +1,114 @@
+from functools import partial, lru_cache
+from typing import Callable, Any, Tuple
+
+import numpy as np
+
+import jax
+from jax import numpy as jnp
+from jax import tree_map
+from netket import jax as nkjax
+from netket.stats import statistics, Stats
+from netket.utils.types import PyTree
+from netket.utils.dispatch import dispatch, TrueT, FalseT
+from netket.utils import mpi
+
+from netket.operator import (
+    DiscreteOperator,
+    AbstractSuperOperator,
+    Squared,
+)
+
+from .exact_state import ExactState
+
+
+def _check_hilbert(A, B):
+    if A.hilbert != B.hilbert:
+        raise NotImplementedError(  # pragma: no cover
+            f"Non matching hilbert spaces {A.hilbert} and {B.hilbert}"
+        )
+
+
+@lru_cache(5)
```

Comment:
```suggestion
# TODO: This cache is here so that we don't re-compute the sparse representation of the operators at every VMC step
# but instead we cache the last 5 used. Should investigate a better way to implement this caching.
@lru_cache(5)
```